homeassistant/components/switchbot/__init__.py

```diff
@@ -13,73 +25,88 @@
     DOMAIN,
 )
-from .coordinator import SwitchbotDataUpdateCoordinator
+from .coordinator import SwitchbotCoordinator
 
 PLATFORMS_BY_TYPE = {
     ATTR_BOT: [Platform.SWITCH, Platform.SENSOR],
     ATTR_CURTAIN: [Platform.COVER, Platform.BINARY_SENSOR, Platform.SENSOR],
     ATTR_HYGROMETER: [Platform.SENSOR],
 }
+CLASS_BY_DEVICE = {
+    ATTR_CURTAIN: switchbot.SwitchbotCurtain,
+    ATTR_BOT: switchbot.Switchbot,
+}
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Switchbot from a config entry."""
     hass.data.setdefault(DOMAIN, {})
+    domain_data = hass.data[DOMAIN]
+
+    if CONF_MAC in entry.data:
+        # Bleak uses addresses not mac addresses which are are actually
+        # UUIDs on some platforms (MacOS).
+        mac = entry.data[CONF_MAC]
+        if "-" not in mac:
+            mac = dr.format_mac(mac)
+        hass.config_entries.async_update_entry(
+            entry,
+            data={k: v for k, v in entry.data.items() if k != CONF_MAC}
+            | {CONF_ADDRESS: mac},
+        )
 
     if not entry.options:
-        options = {
-            CONF_TIME_BETWEEN_UPDATE_COMMAND: DEFAULT_TIME_BETWEEN_UPDATE_COMMAND,
-            CONF_RETRY_COUNT: DEFAULT_RETRY_COUNT,
-            CONF_RETRY_TIMEOUT: DEFAULT_RETRY_TIMEOUT,
-            CONF_SCAN_TIMEOUT: DEFAULT_SCAN_TIMEOUT,
-        }
-
-        hass.config_entries.async_update_entry(entry, options=options)
-
-    # Use same coordinator instance for all entities.
-    # Uses BTLE advertisement data, all Switchbot devices in range is stored here.
-    if DATA_COORDINATOR not in hass.data[DOMAIN]:
-
-        if COMMON_OPTIONS not in hass.data[DOMAIN]:
-            hass.data[DOMAIN][COMMON_OPTIONS] = {**entry.options}
-
-        switchbot.DEFAULT_RETRY_TIMEOUT = hass.data[DOMAIN][COMMON_OPTIONS][
-            CONF_RETRY_TIMEOUT
-        ]
-
-        # Store api in coordinator.
-        coordinator = SwitchbotDataUpdateCoordinator(
-            hass,
-            update_interval=hass.data[DOMAIN][COMMON_OPTIONS][
-                CONF_TIME_BETWEEN_UPDATE_COMMAND
-            ],
-            api=switchbot,
-            retry_count=hass.data[DOMAIN][COMMON_OPTIONS][CONF_RETRY_COUNT],
-            scan_timeout=hass.data[DOMAIN][COMMON_OPTIONS][CONF_SCAN_TIMEOUT],
+        hass.config_entries.async_update_entry(
+            entry,
+            options={
+                CONF_RETRY_COUNT: DEFAULT_RETRY_COUNT,
+                CONF_RETRY_TIMEOUT: DEFAULT_RETRY_TIMEOUT,
+            },
         )
 
-        hass.data[DOMAIN][DATA_COORDINATOR] = coordinator
-
-    else:
-        coordinator = hass.data[DOMAIN][DATA_COORDINATOR]
+    if DEPRECATED_CONFIG_OPTIONS.intersection(entry.options):
+        hass.config_entries.async_update_entry(
+            entry,
+            options={
+                k: v
+                for k, v in entry.options.items()
+                if k not in DEPRECATED_CONFIG_OPTIONS
```

Comment:
Try to keep the deprecated data.

Reply:
5a2ed118342b0c49dcc256ee67b6b7ff9caa17c4